aptos-move/aptos-vm/src/sharded_block_executor/local_executor_shard.rs

```diff
@@ -0,0 +1,249 @@
+    ) -> LocalExecutorClient<S> {
+        let num_threads = num_threads
+            .unwrap_or_else(|| (num_cpus::get() as f64 / num_shards as f64).ceil() as usize);
+        let mut cross_shard_msg_txs = vec![];
+        let mut cross_shard_msg_rxs = vec![];
+        let mut command_txs = vec![];
+        let mut command_rxs = vec![];
+        let mut result_txs = vec![];
+        let mut result_rxs = vec![];
+        for _ in 0..num_shards {
+            let (command_tx, command_rx) = unbounded();
+            let (result_tx, result_rx) = unbounded();
+            command_rxs.push(command_rx);
+            result_txs.push(result_tx);
+            command_txs.push(command_tx);
+            result_rxs.push(result_rx);
+            let mut current_shard_msg_txs = vec![];
+            let mut current_shard_msg_rxs = vec![];
+            for _ in 0..MAX_ALLOWED_PARTITIONING_ROUNDS {
+                let (messages_tx, messages_rx) = unbounded();
+                current_shard_msg_txs.push(messages_tx);
+                current_shard_msg_rxs.push(messages_rx);
+            }
+            cross_shard_msg_txs.push(current_shard_msg_txs);
+            cross_shard_msg_rxs.push(current_shard_msg_rxs);
+        }
+        let executor_shards = command_rxs
+            .into_iter()
+            .zip(result_txs.into_iter())
+            .zip(cross_shard_msg_rxs.into_iter())
+            .enumerate()
+            .map(|(shard_id, ((command_rx, result_tx), cross_shard_rxs))| {
+                Self::new(
+                    shard_id as ShardId,
+                    num_shards,
+                    num_threads,
+                    command_rx,
+                    result_tx,
+                    cross_shard_msg_txs.clone(),
+                    cross_shard_rxs,
+                )
+            })
+            .collect();
+        LocalExecutorClient::new(command_txs, result_rxs, executor_shards)
+    }
+}
+
+pub struct LocalExecutorClient<S: StateView + Sync + Send + 'static> {
+    // Channels to send execute block commands to the executor shards.
+    command_txs: Vec<Sender<ExecutorShardCommand<S>>>,
+    // Channels to receive execution results from the executor shards.
+    result_rxs: Vec<Receiver<Result<Vec<Vec<TransactionOutput>>, VMStatus>>>,
+
+    executor_services: Vec<LocalExecutorService<S>>,
+}
+
+impl<S: StateView + Sync + Send + 'static> LocalExecutorClient<S> {
+    pub fn new(
+        command_tx: Vec<Sender<ExecutorShardCommand<S>>>,
+        result_rx: Vec<Receiver<Result<Vec<Vec<TransactionOutput>>, VMStatus>>>,
+        executor_shards: Vec<LocalExecutorService<S>>,
+    ) -> Self {
+        Self {
+            command_txs: command_tx,
+            result_rxs: result_rx,
+            executor_services: executor_shards,
+        }
+    }
+}
+
+impl<S: StateView + Sync + Send + 'static> ExecutorClient<S> for LocalExecutorClient<S> {
+    fn num_shards(&self) -> usize {
+        self.command_txs.len()
+    }
+
+    fn execute_block(
```

Comment:
Just to confirm my understanding here: right now it's the whole block, but if we do e.g. something along the lines of @xosmig's ideas and send subblocks while producing them, then this intermediate abstraction would be the one to have to change, is that right?

Reply:
Yes, that is correct. If we want the partitioning to be in a streaming fashion then this interface changes, shouldn't be a big change though.